_posts/2023-06-23-run-and-debug-rust-verification-harnesses-natively-on-vscode.md

```diff
@@ -0,0 +1,174 @@
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/run-kani-harness.gif" style="box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);" alt="run-kani-harness" />
+
+Kani can help you generate unit tests containing the counter-example (or values for which the assertion fails).
+Each unit test provides inputs that will either trigger a property failure or satisfy a cover statement.
+This feature, called [concrete playback](https://model-checking.github.io/kani-verifier-blog/2022/09/22/internship-projects-2022-concrete-playback.html), allows you to generate unit tests that call a function with the exact arguments that caused the assertion violation, and the VSCode extension makes using concrete playback easy.
+You can read more about concrete playback in our [documentation](https://model-checking.github.io/kani/debugging-verification-failures.html).
+
+#### Generate a counter-example unit test
+
+Next, we’ll generate the unit test by clicking on the `Run Concrete Playback for stretched_rectangle_can_hold_original` link that appears through a blue link on the error banner.
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/generate-counter-example.gif" style="box-shadow: 0 2px 4px rgba(0, 0, 0, 0.5);" alt="generate-unit-test" />
+
+By simply clicking on a link, we have our counter example unit test pasted directly below the harness. This is what the unit test looks like:
+
+```rust
+#[test]
+fn kani_concrete_playback_stretched_rectangle_can_hold_original() {
+    let concrete_vals: Vec<Vec<u8>> = vec![
+        // 4611686018427387904ul
+        vec![0, 0, 0, 0, 0, 0, 0, 64],
+        // 0ul
+        vec![0, 0, 0, 0, 0, 0, 0, 0],
+        // 2ul
+        vec![2, 0, 0, 0, 0, 0, 0, 0],
+    ];
+    kani::concrete_playback_run(concrete_vals, stretched_rectangle_can_hold_original);
+}
+```
+
+You can see in the gif above that the source is now annotated with two options on top of the generated unit test called  `Run Test (Kani) | Debug Test (Kani)` which allow you to run and debug the test just like any other Rust unit test.
+
+#### Run Kani-generated test
+
+Running the unit test using the Run Test (Kani) button, shows us what we’re expecting–that the current unit test is failing.
+This is because the unit test is using the counter-example to invoke the function `stretched_rectangle_can_hold_original`.
+
+<img src="{{site.baseurl | prepend: site.url}}/assets/images/vs-code-images/run-concrete-playback-test.png" alt="run-concrete-playback-test" />
+
+### Debug Kani unit test
+
+In order to peek under the hood to find out the missing assumptions that lead to unexpected behavior, it is really important to look at the concrete counter examples for which our assertions fail.
+By setting breakpoints and clicking the debug test (Kani) button, you are taken into the debugger which allows you to look at the specific values for which the assertion fails.
```

Comment:
```suggestion
By setting breakpoints and clicking the `Debug Test (Kani)` button, you are taken into the debugger which allows you to inspect the specific values for which the assertion fails.
```

Reply:
What makes debuggers interesting is that you can inspect the values at different points. They aren't necessarily the values for which the assertion fails, those may be computed (somehow) later.